Custom websocket

MarhiievHE opened the pull request:
https://github.com/metarhia/metacom/issues/476

- [x] tests and linter show no problems (`npm t`)
- [x] tests are added/updated for bug fixes and new features
- [x] code is properly formatted (`npm run fix`)
- [ ] description of changes is added in CHANGELOG.md
- [x] update .d.ts typings

MarhiievHE committed:
Add websocket lib implementation

MarhiievHE committed:
Add websocket lib tests

MarhiievHE committed:
Fix 'pong'

tshemsedinov commented on test/websocket/utils/protocolClient.js:
```diff
@@ -0,0 +1,351 @@
+    this.socket.end();
+  }
+
+  on(event, cb) {
+    if (event === 'open') this._openCb = cb;
+    else if (event === 'message') this._messageCb = (buf) => cb(buf);
+    else if (event === 'frame') this._frameCb = cb;
+    else if (event === 'ping') this._pingCb = (buf) => cb(buf);
+    else if (event === 'pong') this._pongCb = (buf) => cb(buf);
+    else if (event === 'close') this._closeCb = cb;
+  }
```
We can just use EventEmitter

MarhiievHE replied:
Yes, that’s a good idea — much cleaner than mocking everything. Done

tshemsedinov commented on test/websocket/utils/protocolClient.js:
```diff
@@ -0,0 +1,351 @@
+// Methods: send(data), sendFrame(opcode, payload, { fin=true, mask=true }),
+// sendText, sendBinary, ping, close
+class ProtocolClient {
+  constructor(url) {
```
Wery long constructor, maybe we can decompose, for example, extracting collection:
`{ [opcode]: [handler] }`

tshemsedinov commented on lib/websocket/utils/result.js:
```diff
@@ -0,0 +1,35 @@
+'use strict';
```
Should be moved to metautil

MarhiievHE replied:
Should I move the Result class to metautil, or do you think it would be better to introduce an algebraic data types?

MarhiievHE committed:
Update server constructor to be compatible with ws

MarhiievHE committed:
Refactor close frames

MarhiievHE committed:
Fix code style

MarhiievHE commented on lib/server.js:
```diff
@@ -197,7 +197,7 @@ class Server {
 
     if (balancer) return;
-    this.wsServer = new ws.Server({ server: this.httpServer });
+    this.wsServer = new WebsocketServer(this.httpServer);
```
Done

MarhiievHE commented on lib/websocket/frame.js:
```diff
@@ -0,0 +1,207 @@
+
+module.exports = {
+  Frame,
+};
```
And it's done

MarhiievHE committed:
Refactor to use EventEmitter

MarhiievHE committed:
Decompose constructor

MarhiievHE committed:
Remove unnecessary conversion to string

MarhiievHE committed:
Remove unnecessary return statement

omelchenko reviewed (commented):
/* not relevant comment */

tshemsedinov committed:
Merge branch 'master' into custom_websocket

tshemsedinov committed:
Remove dead code

tshemsedinov committed:
Websocket close handling, limits, and error cleanup

tshemsedinov committed:
Use `new Frame` instead of `new this`

tshemsedinov committed:
Restrict metacom websocket upgrades to allowed paths

tshemsedinov committed:
Reorganize submodule structure

tshemsedinov committed:
Reorganize constants

tshemsedinov committed:
Remove dead code from websocket frame and connection

- Drop unused normalClose, CLOSE_FRAMES templates, isValid*Frame getters
- Drop redundant DATA_OPCODES guard
- Align types and unit tests with actual APIs

tshemsedinov committed:
Build websocket error close frames at runtime

- Build frames at runtime instead of precomputed buffer
- Replace CLOSE_FRAMES wire bytes with reason maps and Frame.close().toBuffer()

tshemsedinov committed:
Extract #writeFrame helper in Connection send methods

Consolidate duplicated cork/write/uncork and
client masking logic shared by
sendText, sendBinary, sendPing, and sendPong

tshemsedinov committed:
Strengthen websocket frame unit tests for masking and FIN bit

Assert masked state, mask length, and parsed fin flag so frame encoding
bugs are caught instead of passing by accident.

tshemsedinov committed:
Use intermediate variables for errors and other things

tshemsedinov committed:
Use constants for server ping/pong

tshemsedinov committed:
Move all external typings to metacom.d.ts

tshemsedinov committed:
Remove dependency: ws

tshemsedinov committed:
Drop node.js 18, 19, 20, 21 support

tshemsedinov commented:
Landed in: 15453f17 253d0cc 151747b ea7ac1c 93be819 aef43b1 e96e9d8 1203bc2 22d4e59 c835377 980d42a edba5b5 a2515e6 f6c1b83 6d87c36 ecc0c82 de8bced ef19721 7a654ac 8029a66 c8d50ee aa598b2 cc4988e 6980810 8638369 51bf33d
